crates/wasm-mutate/src/mutators/peephole/swap_commutative.rs

```diff
@@ -0,0 +1,176 @@
+            | Operator::F64Add | Operator::F64Mul // Check for float incosistency
+            // TODO do the others
+            => {
+                true
+            }
+            _ => {
+                false
+            }
+        }
+    }
+
+    fn get_operator_type(&self, op: &Operator) -> Result<ValType> {
+        match op {
+            Operator::I32Add
+            | Operator::I32Mul
+            | Operator::I32Or
+            | Operator::I32And
+            | Operator::I32Xor => Ok(ValType::I32),
+            Operator::I64Add
+            | Operator::I64Mul
+            | Operator::I64And
+            | Operator::I64Or
+            | Operator::I64Xor => Ok(ValType::I64),
+            Operator::F32Add | Operator::F32Mul => Ok(ValType::F32),
+            Operator::F64Add | Operator::F64Mul => Ok(ValType::F64),
+            // TODO do the others
+            _ => Err(Error::UnsupportedType(EitherType::Operator(format!(
+                "{:?}",
+                op
+            )))),
+        }
+    }
+}
+
+impl CodeMutator for SwapCommutativeOperator {
+    fn mutate(
+        &self,
+        _: &WasmMutate,
+        _: &mut SmallRng,
+        funcreader: &mut FunctionBody,
+        operator_index: usize,
+        input_wasm: &[u8],
+    ) -> Result<Function> {
+        let opreader = funcreader.get_operators_reader().unwrap();
+        let body_range = opreader.range();
+        let opreaderiterator = opreader.into_iter_with_offsets();
+        let operators = opreaderiterator
+            .collect::<wasmparser::Result<Vec<(Operator, usize)>>>()
+            .unwrap();
+
+        let mut localreader = funcreader.get_locals_reader().unwrap();
+        // Get current locals and map to encoder types
+        let mut local_count = 0;
+        let mut current_locals = (0..localreader.get_count())
+            .map(|f| {
+                let (count, ty) = localreader.read().unwrap();
+                local_count += count;
+                (count, map_type(ty).unwrap())
+            })
+            .collect::<Vec<(u32, ValType)>>();
+
+        // add two temporary locals, the last two
+        let (operator, _) = &operators[operator_index];
+        current_locals.push((2, self.get_operator_type(operator)?));
+
+        let mut newf = Function::new(current_locals);
+        let mut idx = 0;
+
+        let mut newoffset = 0;
+        for (_, offset) in operators {
+            newoffset = offset;
+            if idx == operator_index {
+                // Copy previous code to the body
+                let previous = &input_wasm[body_range.start..offset];
+                newf.raw(previous.iter().copied());
+                // Inject new code to swap operands
+                newf.instruction(Instruction::LocalTee(local_count + 1));
+                newf.instruction(Instruction::LocalTee(local_count + 2));
+                break; // this break allows to copy the remaining buffer of the current reader
```

Comment:
I think `local.tee`s aren't quite right here, and will result in the following sequence, no?

```wat
...         // Stack: [a b] ; Locals: x = 0, y = 0
local.tee x // Stack: [a b] ; Locals: x = b, y = 0
local.tee y // Stack: [a b] ; Locals: x = b, y = b
```

Instead I think we want to insert this snippet:

```wat
...         // Stack: [a b] ; Locals: x = 0, y = 0
local.set x // Stack: [a]   ; Locals: x = b, y = 0
local.set y // Stack: []    ; Locals: x = b, y = a
local.get x // Stack: [b]   ; Locals: x = b, y = a
local.get y // Stack: [b a] ; Locals: x = b, y = a
```

Reply:
Done